### ISSUE TITLE
PyTorch multiprocessing using single CPU core

### ISSUE THREAD

**Issue by xdever** [pytorch/pytorch#8126]
## Issue description

I was trying to load data with DataLoader with multiple workers, and I noticed that although it creates the processes they all run on the same CPU core, thus the data loading is very slow.

I succeeded in creating a minimal example (below). If I run the script below, it uses 4 CPU cores. If I uncomment the "import torch" line, only single core is used. If I also uncomment the taskset command, it will use 4 cores again.

## Code example

```python
#!/usr/bin/env python3

#import torch
import os
def loop():
    while True:
        pass

#os.system("taskset -p 0xff %d" % os.getpid())
from multiprocessing import Process
for i in range(4):
    Process(target=loop).start()
```

## System Info
Although the collect_env.py prints gcc 8.1.0, it was overridden to 7.3.1 during build. Also cudnn detection failed, its version is 7.1.3.

Built from source (commit f24d715e235eb7d188cb610d7b29386b9eaf0ab9), command: "CC=gcc-7 CXX=g++-7 python setup.py bdist_wheel", package installed with pip install --user.

PyTorch version: 0.5.0a0+f24d715
Is debug build: No
CUDA used to build PyTorch: 9.2.88

OS: Antergos Linux
GCC version: (GCC) 8.1.0
CMake version: version 3.11.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration: 
GPU 0: TITAN X (Pascal)
GPU 1: GeForce GTX 1070

Nvidia driver version: 396.24
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy (1.13.1)
[pip3] torch (0.5.0a0+f24d715)
[pip3] torchfile (0.1.0)
[pip3] torchvision (0.1.8)
[conda] Could not collect

**Comment by nlgranger** [comment]
Pytorch ships its [own patched version of multiprocessing](https://pytorch.org/docs/master/notes/multiprocessing.html) which they say is a drop-in replacement. Have you tried using it?

**Comment by xdever** [comment]
I noticed the problem while I was using the built-in DataLoader class, which uses pytorch's internal multiprocessing. I just used pythons multiprocessing in the example to demonstrate that the whole program will become locked to one CPU core when pytorch is imported. I also tried explicitly changing "from multiprocessing import Process" to "from torch.multiprocessing import Process" and exactly the same thing happens.

**Comment by bhushan23** [comment]
@xdever, I am unable to repro this issue. I tried similar versions. Only difference was I am using CPU build.
Can you try CPU build at your end? This might confirm root cause

**Comment by xdever** [comment]
Tried a CPU build (CC=gcc-7 CXX=g++-7 NO_CUDA=1 python setup.py bdist_wheel), it doesn't help. Also tried to build it with gcc-8. I tried the newest master (147fc6b9cc2c8c4e2c504156825a5e3c0e3a4e8c) with the same result. I also ran a full Linux update, still nothing changed.

![screenshot](https://user-images.githubusercontent.com/2706617/41146266-1841752e-6b03-11e8-9ec5-d1ba607cd315.png)

**Comment by BenediktKersjes** [comment]
Where are having the same issue. Is there a workaround to make the DataLoader use all CPU cores?

**Comment by xdever** [comment]
I worked it around by putting "os.system("taskset -p 0xff %d" % os.getpid())" in my script.

**Comment by Kylin9511** [comment]
@xdever I would like to thank you for your solution `os.system("taskset -p 0xff %d" % os.getpid())`

That really works for me. And all the rest solutions including `torch.set_num_threads(20)` and setting `OMP_NUM_THREADS` do not make any different...

And some more information for those brwosing this issue:
> `taskset -p XXX PID` is a linux command, that can change the CPU affinity of certain process (assgined by the PID).
> `XXX` should be an Hexadecimal integer and it determines which CPUs to use for the process. For instance, `0xff` is `1111 1111` in binary, which will activate the first 8 CPUs. If you want to use the  CPU 2, 4, 6 and 8, you can set `XXX` to `0x55`(`01010101`). `0xffffffff` for all the fist 32 CPUs of course.

**Comment by seebag** [comment]
I got the same problem. Pytorch seems to mask the CPU affinity by default and subsequent forks are forced to used only one cpu core.

Here is a simple script to reproduce the issue : 

```
import multiprocessing
import psutil

def target():
    print("In target = ", len(psutil.Process().cpu_affinity()), " cores available")

print("### Without torch")
p = multiprocessing.Process(target=target)
p.start()
print("Before join = ", len(psutil.Process().cpu_affinity()), " cores available")
p.join()
p.terminate()

print("### With torch")
import torch
p = multiprocessing.Process(target=target)
p.start()
print("Before join=", len(psutil.Process().cpu_affinity()), " cores available")
p.join()
p.terminate()
```

With the following result : 
```
### Without torch
Before join = 32  cores available
In target =  32  cores available
### With torch
Before join = 32  cores available
In target =  1  cores available
```

_Note that if cpu cores are hyperthreaded, you could got 2 cores available on the last line_

The prevent such problem, one could add `psutil.Process().cpu_affinity([])` in the target function.

Any idea on what happen inside pytorch ?
